chain/client/src/view_client.rs

```diff
@@ -97,6 +97,8 @@ pub struct ViewClientActor {
     network_adapter: Arc<dyn NetworkAdapter>,
     pub config: ClientConfig,
     request_manager: Arc<RwLock<ViewClientRequestManager>>,
+    state_header_request_cache: Arc<Mutex<TimedCache<(CryptoHash, ShardId), ()>>>,
```

Comment:
I don't see how this would help. I saw a situation where we were sending to peer 11k messages of size 2mb each over period of 30m for a node with 32gb ram. So that would be less than 7m for a node with 8 gb ram.

I don't have stat's on how many of those requests are unique so this may not help at all. Even if we slow down the amount the grown of data by factor of 10x, the node will die in an hour instead of 7m, so likely we won't have much affect at all. Though by looking at rate of requests, we see 6 per seconds, so either there are a lot of duplicates, or those are requests to different peers for the same CryptoHash.

Reply:
> so either there are a lot of duplicates

That is very likely the case given the suboptimal state sync design (#3777). Also this protects against spams

Reply:
Can you provide us with stats of what percentage of requests are duplicates?

Reply:
I don't have any stats, but the total number of state parts should be around ~3k on testnet. cc @mikhailOK

Reply:
With 3k parts and 30 second throttle per part we can still get 100 requests/second, this seems like a problem

Reply:
On one node that ran out of memory where we send 11k StateRequests to one shard I had a print which shows time, length of message send to given peer. For example:
```

[2mFeb 24 22:07:49.761[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 886107
[2mFeb 24 22:07:49.762[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1029446
[2mFeb 24 22:07:49.763[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 4112816
[2mFeb 24 22:07:49.765[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1891264
[2mFeb 24 22:07:49.765[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 695405
[2mFeb 24 22:07:49.766[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 764465
[2mFeb 24 22:07:49.767[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3754816
[2mFeb 24 22:07:49.769[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 2124892
[2mFeb 24 22:07:49.769[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 924100
[2mFeb 24 22:07:49.770[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1905068
[2mFeb 24 22:07:49.771[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1814682
```

Here is the total number of requests:
```bash
> cat xxx.txt | grep StateResponse | wc -l
 11464
```

If the group requests by the minute, and length, and print unique entries (that will be an upper bound)
```bash
> cat xxx.txt | grep StateResponse | cut -c 13-17,100- | sort | uniq-c | wc -l
6404
```

Therefore the solution suggested in in request would reduce the number of requests by up to 45%, so this would not have much of an effect on the issue we were trying to solve.

Reply:
We do see that the same chunks were requested multiple times over period of 30 minutes, here is an example for chunk of size `3661359`.
```
[2mFeb 24 21:38:28.101[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:38:42.753[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:39:19.962[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:39:42.534[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:39:57.039[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:41:26.905[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:41:44.142[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:42:55.326[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:42:55.401[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:43:36.778[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:43:42.491[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:44:31.613[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:45:35.039[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:46:21.572[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:47:10.639[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:47:33.760[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:50:14.535[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:50:19.451[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:50:36.836[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:52:35.851[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:55:17.819[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:55:20.559[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:55:35.185[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:56:22.179[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:56:27.402[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:57:14.904[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:57:15.551[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:58:19.494[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:58:29.978[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 21:59:33.337[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:00:40.557[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:00:41.837[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:00:42.152[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:01:25.373[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:02:48.739[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:03:26.730[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:03:38.884[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:03:46.039[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:04:44.783[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:05:27.521[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:05:28.175[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:07:39.217[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
[2mFeb 24 22:07:49.780[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3661359
```
We do see that we were trying to send the same StateResponse multiple times, usually every 30s, there were some duplicates, so in theory we could reduce the traffic in half.

Reply:
If we were to send each chunk to given peer at most once in period of 30m, we would reduce the number of chunks send to
```
> cat xxx.txt | grep StateResponse | cut -c 100- | sort | uniq-c | wc -l
353
```

Reply:
@mikhailOK I updated the implementation to serve at most 30 requests every 30s.